apps/server/src/shared/domain/service/file-copy-append.service.ts

```diff
@@ -72,23 +76,151 @@ export class FileCopyAppendService {
 	}
 
-	private replaceFileGroup(elements: CopyStatus[] = [], fileGroupStatus: CopyStatus): CopyStatus[] {
-		return elements.map((el) => (el.type === CopyElementType.FILE_GROUP ? fileGroupStatus : el));
+	private setFileGroupStatus(elements: CopyStatus[] = [], updatedFileGroupStatus: CopyStatus): CopyStatus[] {
+		const fileGroupStatus = this.getFileGroupStatus(elements);
+		if (fileGroupStatus) {
+			return elements.map((el) => (el.type === CopyElementType.FILE_GROUP ? updatedFileGroupStatus : el));
+		}
+		return [...elements, updatedFileGroupStatus];
 	}
 
 	private createFileStatuses(files: FileDto[]): CopyStatus[] {
 		return files.map((file) => ({
 			type: CopyElementType.FILE,
-			name: file.name,
+			title: file.name,
 			status: CopyStatusEnum.SUCCESS,
 		}));
 	}
+
+	private createFileStatusesByCopyResult(files: FileLegacyResponse[]): CopyStatus[] {
+		return files.map(({ oldFileId, fileId, filename }) => ({
+			type: CopyElementType.FILE,
+			status: fileId ? CopyStatusEnum.SUCCESS : CopyStatusEnum.FAIL,
+			title: filename ?? `(old fileid: ${oldFileId})`,
+		}));
+	}
+
+	async copyFiles(copyStatus: CopyStatus, courseId: EntityId, userId: EntityId, jwt: string): Promise<CopyStatus> {
+		if (copyStatus.type === CopyElementType.LESSON) {
+			return this.copyEmbeddedFilesOfLessons(copyStatus, courseId, userId);
+		}
+		if (copyStatus.type === CopyElementType.TASK) {
+			const updatedStatus = await this.appendFilesToTask(copyStatus, jwt);
+			return this.copyEmbeddedFilesOfTasks(updatedStatus, courseId, userId);
+		}
+		if (copyStatus.elements && copyStatus.elements.length > 0) {
+			copyStatus.elements = await Promise.all(
+				copyStatus.elements.map((el) => this.copyFiles(el, courseId, userId, jwt))
+			);
+			copyStatus.status = this.copyHelperService.deriveStatusFromElements(copyStatus.elements);
+		}
+		return Promise.resolve(copyStatus);
+	}
+
+	async copyEmbeddedFilesOfLessons(
+		lessonCopyStatus: CopyStatus,
+		courseId: EntityId,
+		userId: EntityId
+	): Promise<CopyStatus> {
+		const lesson = lessonCopyStatus.copyEntity as Lesson;
+		const legacyFileIds: string[] = [];
+
+		lesson.contents.forEach((item: IComponentProperties) => {
+			if (item.content === undefined || !('text' in item.content)) {
+				return;
+			}
+
+			const contentFileIds = this.extractOldFileIds(item.content.text);
+
+			legacyFileIds.push(...contentFileIds);
+		});
+
+		if (legacyFileIds.length > 0) {
+			const fileCopyResults: FileLegacyResponse[] = await Promise.all(
+				legacyFileIds.map((fileId) => this.fileLegacyService.copyFile({ fileId, targetCourseId: courseId, userId }))
+			);
+
+			fileCopyResults.forEach(({ oldFileId, fileId, filename }) => {
+				lesson.contents = lesson.contents.map((item: IComponentProperties) => {
+					if ('text' in item.content && fileId && filename) {
+						const text = this.replaceOldFileUrls(item.content.text, oldFileId, fileId, filename);
+						return { ...item, content: { ...item.content, text } };
```

Comment:
To put the return value to a function before like
```
const abc = value

return abc
```

force you to think about how the return value should be named. The name as description. It is easer to read and understand the code and helps to avoid misstakes by return values.

Reply:
fixed